### PR TITLE
Triple.Extra

### LINKED ISSUE #21: Triple library missing
One of the types where an -extra library would be handy are triples `(a, b, c)`.

Does it make sense to add such a library? Should it go into `Tuple.Extra` or should we make a new name for it? `Triple.Extra`?

Some candidate functions:

- `triple : a -> b -> c -> (a, b, c)`
- `first : (a, b, c) -> a`
- `second : (a, b, c) -> b`
- `third : (a, b, c) -> c`
- `mapFirst`, `mapSecond`, `mapThird`
- `map : (a -> b) -> (a, a, a) -> (b, b, b)`
- `mapAll : (a -> x) -> (b -> y) -> (c -> z) -> (a, b, c) -> (x, y, z)`
- `from : a -> (a, a, a)`
- and perhaps analogues of `Tuple.Extra` functions, although perhaps we could wait for demand there...

### PR DESCRIPTION
**Pull request by gampleman**
This adds a library for dealing with Triples (fixes #21).

I think the most controversial idea here are the swizzles. I'm happy to delete them for now if the other maintainers have qualms.

### CODE REVIEW

**Review by lue-bird — COMMENTED** [review]
I can't say much because I've never used triples
  - would be nice to still add the codegen in case one might want to change the documentation in the future for example
  - the swizzles do in fact seem only very rarely useful and I'd always prefer explicit construction personally
  - `fromListHead`: maybe rename it to `fromListBeginning`?
  - `joinBy` seems somewhat inconsisten because for one we don't have `sumBy`, `productBy` etc either and for two in all other `-By` operations like `minimumBy` or `sortBy` the given function(s) do not map the result which means a name like `joinMap` is probably more consistent and for three I feel like the arguments in the explicit `join sep (map a b c triple)` are much more clear

**Comment by gampleman** [comment]
Hm I tend to agree with some of these sentiments. The functions I added are just stuff we have in Tuple.Extra. But perhaps it's better to be more conservative to begin with.

I'll probably like to keep the sorting functions (or at least `sortWith`) since they're somewhat challenging to implement correctly.

**Comment by lydell** [comment]
I’m curious about the swizzles. Is that something you came up with yourself, or something you’ve seen somewhere else? Also just for curiosity, what could they be used for?

**Comment by gampleman** [comment]
> I’m curious about the swizzles. Is that something you came up with yourself, or something you’ve seen somewhere else? Also just for curiosity, what could they be used for?

They're a common feature of GPU oriented languages like GLSL. I always found them super fun, but perhaps the reason they'd be useful in those contexts is that those languages generally compute around fix length vectors, whereas actually triples are fairly rare in Elm (and we don't actually particularly want to promote them).

So you can do things like:

```glsl

vec4 opaque = vec4(color.rgb, 1.0);

vec3 mirrored = position.yxz;

vec3 onlyRed = color.rrr;
```

and so on

File: src/Triple/Extra.elm [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        --> Just ( 1, 2, 3 )
+
+    Triple.Extra.fromList [ 1, 2 ]
+        --> Nothing
+
+-}
+fromList : List a -> Maybe ( a, a, a )
+fromList lst =
+    case lst of
+        a :: b :: c :: _ ->
```
Comment (lue-bird) [inline src/Triple/Extra.elm:252]: From the name alone, I would expect this definition
```elm
case lst of
    [ a, b, c ] -> Just ( a, b, c )
    _ -> Nothing
```
so I'd still prefer a name that makes it clear the elements are from the beginning. All other changes look good!

Comment (gampleman) [reply src/Triple/Extra.elm:252]: I originally thought of that definition too. But it seemed to me that this definition would be more versatile (although perhaps not - it's a fairly simple fix to do `List.take 3 >> Triple.Extra.fromList`).

Should we then just have the strict version?

Comment (lue-bird) [reply src/Triple/Extra.elm:252]: Well I couldn't tell you a use case for either that wouldn't be nicer as an explicit case of.
But if we decide to keep either, I'm in favor of the `fromListBeginning` taking the first 3

Comment (gampleman) [reply src/Triple/Extra.elm:252]: OK, let's punt on this one and just remove it for now.

**Review by lue-bird — APPROVED** [review]

**Pull request merged**